server/src/main/java/org/elasticsearch/action/fieldcaps/RequestDispatcher.java

```diff
@@ -255,10 +268,34 @@ private static class IndexSelector {
         private final Set<ShardId> unmatchedShardIds = new HashSet<>();
         private final Map<ShardId, Exception> failures = new HashMap<>();
 
-        IndexSelector(List<ShardIterator> shardIts) {
+        IndexSelector(
+            String clusterAlias,
+            List<ShardIterator> shardIts,
+            QueryBuilder indexFilter,
+            long nowInMillis,
+            CoordinatorRewriteContextProvider coordinatorRewriteContextProvider
+        ) {
             for (ShardIterator shardIt : shardIts) {
-                for (ShardRouting shard : shardIt) {
-                    nodeToShards.computeIfAbsent(shard.currentNodeId(), node -> new ArrayList<>()).add(shard);
+                boolean canMatch = true;
+                final ShardId shardId = shardIt.shardId();
+                if (indexFilter != null && indexFilter instanceof MatchAllQueryBuilder == false) {
```

Comment:
We can use CanMatchPreFilterSearchPhase to avoid duplicates, but this approach is simpler.